stetho/src/main/java/com/facebook/stetho/inspector/network/DownloadingAsyncPrettyPrinterFactory.java

```diff
@@ -0,0 +1,170 @@
+                "Error while downloading schema for pretty printing: " + e.getMessage());
+          } catch (ExecutionException e) {
+            Throwable cause = e.getCause();
+            ExceptionUtil.propagateIfInstanceOf(cause, IOException.class);
+            throw ExceptionUtil.propagate(cause);
+          }
+        }
+        public DownloadingAsyncPrettyPrinterFactory.PrettyPrinterDisplayType getPrettifiedType() {
+          return result.getDisplayType();
+        }
+      };
+    }
+  }
+
+  /**
+   * Match the correct header that contains information about the schema uri
+   * @param headerName header name of a response that needs to be pretty printed
+   * @param headerValue header value which contains the URI for
+   * the schema data needed to pretty print the response body
+   * @return MatchResult that has the schema uri and the type of prettified result. Null
+   * if there is no correct header match.
+   */
+  @Nullable
+  protected abstract MatchResult matchAndParseHeader(String headerName, String headerValue);
+
+  protected abstract void doPrint(PrintWriter output, InputStream payload, String schema)
+      throws IOException;
+
+  @Nullable
+  private static URL parseURL(String uri) {
+    try {
+      return new URL(uri);
+    } catch (MalformedURLException e) {
+      return null;
+    }
+  }
+
+  private static void doErrorPrint(PrintWriter output, InputStream payload, String errorMessage)
+    throws IOException {
+    output.print(errorMessage + "\n" + Util.readAsUTF8(payload));
+  }
+
+  private static AsyncPrettyPrinter getErrorAsyncPrettyPrinter(
+      final String headerName,
+      final String headerValue) {
+    return new AsyncPrettyPrinter() {
+      @Override
+      public void printTo(PrintWriter output, InputStream payload) throws IOException {
+        String errorMessage = "[Failed to parse header: "
+            + headerName + " : " + headerValue + " ]";
+        doErrorPrint(output, payload, errorMessage);
+      }
+
+      @Override
+      public PrettyPrinterDisplayType getPrettifiedType() {
+        return PrettyPrinterDisplayType.TEXT;
+      }
+    };
+  }
+
+
+  public enum PrettyPrinterDisplayType {
+    JSON("Json"),
+    HTML("Html"),
+    TEXT("Text");
+
+    private final String mDisplayType;
+
+    private PrettyPrinterDisplayType(String displayType) {
+      mDisplayType = displayType;
+    }
+
+    public String getDisplayType() {
+      return mDisplayType;
+    }
+  }
+
+  protected class MatchResult {
+    private String mSchemaUri;
+    private PrettyPrinterDisplayType mDisplayType;
+
+    public MatchResult(String schemaUri, PrettyPrinterDisplayType displayType) {
+      mSchemaUri = schemaUri;
+      mDisplayType = displayType;
+    }
+
+    public String getSchemaUri() {
+      return mSchemaUri;
+    }
+
+    public PrettyPrinterDisplayType getDisplayType() {
+      return mDisplayType;
+    }
+  }
+
+  private class Request implements Callable<String> {
```

Comment:
Can this be a static class?  Classes and functions without side effects should be marked as such to help the reader easily understand their scope